pytensor/scan/rewriting.py

```diff
@@ -1103,6 +1111,71 @@ def sanitize(x):
         return at.as_tensor_variable(x)
 
 
+@node_rewriter([Scan])
+def while_scan_merge_subtensor_last_element(fgraph, scan_node):
```

Comment:
This rewrite should be thought of as a specialization of `local_subtensor_merge` which fuses two subtensor (indexing) operations into a single one. That is potentially useful on it's own, regardless of the driving goal in this PR as a stepping stone to optimize memory of While scans.

The question of why didn't `save_mem_new_scan` work already, was that by default `local_subtensor_merge` fuses two subtensor operations in a way that refers to the shape of of the original tensor: `x[1:][-1] -> x[len(x)-1 if len(x) > 1 else ... :]`. The problem with this for while scans is that is the reliance of `len(x)` which can only be assessed by running the while loop. This is not a problem for normal scans as we can always know `len(x)` from the inputs alone without having to execute it.

The `save_mem_new_scan` aborts if the final expression still references the old scan, as that would mean there would be now two scan loops instead of a smaller one, which would be worse than we started of with.

This PR is a workaround to support the optimization of while scans by handling `x[1:][-1]` specially, as we know that operation is valid as long as there is one single step (hence the Assert).

Reply:
Thank you for the explanation!